### PR TITLE
JAMES-2597 Create a Cassandra + Rabbitmq + Swift + Ldap

### PR DESCRIPTION
**Pull request by hoangdat**
JIRA: https://issues.apache.org/jira/browse/JAMES-2597

CI: https://ci.linagora.com/linagora/lgs/openpaas/james/issues/883

### CODE REVIEW

File: server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml [hunk 1]
```diff
@@ -0,0 +1,175 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml]: You are missing the Apache V2.0 license

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/resources/ldif-files/populate.ldif [hunk 2]
```diff
@@ -0,0 +1,11 @@
+dn: ou=people, dc=james,dc=org
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/resources/ldif-files/populate.ldif]: Given that we import LdapTestExtension from cassandra-ldap-guice mvn module, do we really need to duplicate this LDIF and docker file?

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/resources/ldif-files/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,3 @@
+FROM dinkel/openldap:latest
+
+COPY populate.ldif /etc/ldap/prepopulate/prepop.ldif
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/resources/ldif-files/Dockerfile]: Given that we import LdapTestExtension from cassandra-ldap-guice mvn module, do we really need to duplicate this LDIF and docker file?

**Comment by trantienduchn** [comment]
```
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Errors: 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:230 » ConditionTimeout
[6554af76111893eda011023e2f0c520130958ca8] [INFO] 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Tests run: 655, Failures: 0, Errors: 1, Skipped: 2
```
test this please

File: server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>apache-james-backends-es</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>apache-james-backends-rabbitmq</artifactId>
+            <version>${project.version}</version>
```
Comment (trantienduchn) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml]: do we really need this `<version>` tag

File: server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>james-server-cassandra-rabbitmq-guice</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>james-server-data-ldap-integration-testing</artifactId>
+            <version>${project.version}</version>
```
Comment (trantienduchn) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml]: and here too?

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java [hunk 6]
```diff
@@ -37,54 +38,84 @@
...
+        @Test
+        default void userFromLdapShouldLoginViaImapProtocol(GuiceJamesServer server) throws IOException {
+            IMAPClient imapClient = new IMAPClient();
+            imapClient.connect(JAMES_SERVER_HOST, server.getProbe(ImapGuiceProbe.class).getImapPort());
+
+            assertThat(imapClient.login(JAMES_USER, PASSWORD)).isTrue();
+        }
+    }
+
+    interface ContratSuite extends JmapJamesServerContract, MailsShouldBeWellReceived, UserFromLdapShouldLogin {}
```
Comment (trantienduchn) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: s/`ContratSuite`/`ContractSuite`

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java [hunk 7]
```diff
@@ -37,54 +38,84 @@
...
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class WithSwift implements ContratSuite {
+        @RegisterExtension
+        JamesServerExtension testExtension = new JamesServerExtensionBuilder()
+            .extension(new EmbeddedElasticSearchExtension())
+            .extension(new CassandraExtension())
+            .extension(new RabbitMQExtension())
+            .extension(new LdapTestExtention())
+            .extension(new SwiftBlobStoreExtension())
```
Comment (trantienduchn) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: actually, you still need a binding statement to bind SwiftBlobStore to `BlobStore`, otherwise. James will use CassandraBlobStore as the default. [See](https://github.com/linagora/james-project/blob/735ef66c38f0bf594e420d8b0822979753ccbad7/server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreChoosingModule.java) from PR #1917 
to get the way to deal with this problem. An example is in #1911 : https://github.com/linagora/james-project/blob/3d7b0bb89cd2112ec11fb730b2c7abfe10e9d232/server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQSwiftJmapTestRule.java#L71

**Comment by trantienduchn** [comment]
```
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Errors: 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:230 » ConditionTimeout
[6554af76111893eda011023e2f0c520130958ca8] [INFO] 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Tests run: 655, Failures: 0, Errors: 1, Skipped: 2
```
test this please

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+            assertThat(imapClient.login(JAMES_USER, PASSWORD)).isTrue();
+        }
+    }
+
+    interface ContratSuite extends JmapJamesServerContract, MailsShouldBeWellReceived, UserFromLdapShouldLogin {}
+
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class WithSwift implements ContratSuite {
```
Comment (trantienduchn) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: Probably you are running `WithSwift` and `WithoutSwift` with two different docker bootstraps(cassandra+rabbit+swift+es up and down for executing `WithSwift` and again for `WithoutSwift`). I guess we could use once to save execution time of this test class.

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java (hunk 7 above)
Comment (chibenwa) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: We can bind **the choosing configuration** to the one with Swift

**Comment by trantienduchn** [comment]
```
[6554af76111893eda011023e2f0c520130958ca8] [ERROR]   VacationIntegrationTest.jmapVacationShouldNotSendNotificationTwice:243->assertOneMessageReceived:393 1 expectation failed.
[6554af76111893eda011023e2f0c520130958ca8] JSON path [1][1].list doesn't match.
[6554af76111893eda011023e2f0c520130958ca8] Expected: a collection with size <1>
[6554af76111893eda011023e2f0c520130958ca8]   Actual: [{mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=5, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}, {mailboxIds=[5], from={name=benwa@domain.tld, email=benwa@domain.tld}, textBody=Message explaining my wonderful vacations, id=6, to=[{name=matthieu@domain.tld, email=matthieu@domain.tld}]}]
```
test this please

**Comment by trantienduchn** [comment]
```
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Errors: 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR]   MailQueueContract.concurrentEnqueueDequeueShouldNotFail:387 » NotTerminated
[6554af76111893eda011023e2f0c520130958ca8] [INFO] 
[6554af76111893eda011023e2f0c520130958ca8] [ERROR] Tests run: 183, Failures: 0, Errors: 1, Skipped: 0
```
test this pleasee

**Comment by trantienduchn** [comment]
rebase master, force push

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java (hunk 8 above)
Comment (trantienduchn) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: We're now having `SwiftDockerSingleton` and `CassandraDockerSingleton`, `RabbitMQDockerSingleton` is ahead in the Jmap tests PR, you can wait till it is merged, then apply rabbit singleton to `RabbitMQExtension`. Now you are free to make `LdapDockerSingleton`

**Comment by chibenwa** [comment]
I am a bit confused, as `cassandra-ldap-guice` implement two times `JmapJamesServerContract` and does not implement `JamesServerContract`... (CassandraLdapJamesServerTest implements JmapJamesServerContract while it should implement JamesServerContract)

 - Could you fix this in that project?
 - Could you run JamesServerContract for `cassandra-rabbitmq-ldap-guice` ?

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.james.utils.SpoolerProbe;
+import org.awaitility.Awaitility;
+import org.awaitility.Duration;
+import org.awaitility.core.ConditionFactory;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+class CassandraRabbitMQLdapJamesServerTest {
```
[the remark this answers is not in the document]
Comment (hoangdat) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: I added `JamesServerContract`

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java (hunk 8 above)
Comment (hoangdat) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJamesServerTest.java]: I implemented

File: server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>james-server-testing</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>${junit.jupiter.version}</version>
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/pom.xml]: This version should not be needed

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+import org.awaitility.core.ConditionFactory;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+class CassandraRabbitMQLdapJmapJamesServerTest {
+    private static final int LIMIT_TO_10_MESSAGES = 10;
+    private static final String JAMES_USER = "james-user";
+    private static final String PASSWORD = "secret";
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java]: These constants are directly related to the Dockerfile defined in `server/data/data-ldap-integration-testing`

They are duplicated in `ReadOnlyUsersLDAPRepositoryTest`, `JmapJamesServerContract`, and this class.

I think it is time to factorize this a bit.

I propose to introduce an interface holding these constants in `server/data/data-ldap-integration-testing` so that it can be reused in other places. I propose to put this in the `DockerLdapSingleton` class defined below.

File: server/container/guice/pom.xml [hunk 12]
```diff
@@ -74,6 +75,20 @@
...
+            <dependency>
+                <groupId>${james.groupId}</groupId>
+                <artifactId>james-server-cassandra-ldap-guice</artifactId>
+                <version>${project.version}</version>
+                <type>test-jar</type>
+                <scope>test</scope>
+            </dependency>
+            <dependency>
+                <groupId>${james.groupId}</groupId>
+                <artifactId>james-server-data-ldap-integration-testing</artifactId>
```
Comment (chibenwa) [inline server/container/guice/pom.xml]: This dependency is not related to guice. We should add it in `james-project/pom.xml` **dependencyManagement** section.

File: server/data/data-ldap-integration-testing/pom.xml [hunk 13]
```diff
@@ -38,6 +38,12 @@
             <artifactId>james-server-data-ldap</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>james-server-guice-common</artifactId>
```
Comment (chibenwa) [inline server/data/data-ldap-integration-testing/pom.xml]: Why is this import needed?

Could you check that you can remove it?

File: server/data/data-ldap-integration-testing/src/test/java/org/apache/james/user/ldap/DockerLdapRule.java [hunk 14]
```diff
@@ -17,39 +17,30 @@
  * under the License.                                           *
```
Comment (chibenwa) [inline server/data/data-ldap-integration-testing/src/test/java/org/apache/james/user/ldap/DockerLdapRule.java]: You should not be doing that move. `TestExtensions` are specific to guice, and we should not leak them in other places.

As you don't use it in `data-ldap-integration-test` I bet you could safely move it back to `cassandra-ldap-guice`...

**Comment by chibenwa** [comment]
test this please

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java (hunk 11 above)
Comment (hoangdat) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java]: i pushed in `DockerLdapSingleton`

File: server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java [hunk 15]
```diff
@@ -19,6 +19,9 @@
 
 package org.apache.james;
 
+import static org.apache.james.user.ldap.DockerLdapSingleton.JAMES_USER;
+import static org.apache.james.user.ldap.DockerLdapSingleton.LIMIT_TO_10_MESSAGES;
```
Comment (chibenwa) [inline server/container/guice/cassandra-ldap-guice/src/test/java/org/apache/james/CassandraLdapJamesServerTest.java]: `LIMIT_TO_10_MESSAGES` is a JMAP related constant, it has nothing to do in `DockerLdapSingleton`

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    private static Duration slowPacedPollInterval = ONE_HUNDRED_MILLISECONDS;
+
+    private static ConditionFactory calmlyAwait = Awaitility
+        .with().pollInterval(slowPacedPollInterval)
+        .and().with().pollDelay(slowPacedPollInterval)
+        .await();
+
+    interface MailsShouldBeWellReceived {
```
Comment (mbaechler) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java]: There's nothing specific to Ldap here, right ? Why do we need to write that test ? (I mean, it's probably already written elsewhere)

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    interface ContractSuite extends JmapJamesServerContract, MailsShouldBeWellReceived,
+        UserFromLdapShouldLogin, JamesServerContract {}
+
+    @Nested
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class WithSwift implements ContractSuite {
+        @RegisterExtension
+        JamesServerExtension testExtension = new JamesServerExtensionBuilder()
```
Comment (mbaechler) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java]: can we try to be DRY here : we duplicate all but a single line

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java (hunk 16 above)
Comment (hoangdat) [reply server/container/guice/cassandra-rabbitmq-ldap-guice/src/test/java/org/apache/james/CassandraRabbitMQLdapJmapJamesServerTest.java]: updated my fixup!

**Comment by chibenwa** [comment]
Merged
